File: uhrwerk-common/src/main/java/io/qimia/uhrwerk/common/model/Partition.java
package io.qimia.uhrwerk.common.model;

import com.google.common.base.Objects;
import java.io.Serializable;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import net.openhft.hashing.LongHashFunction;

public class Partition extends BaseModel implements Comparable<Partition>, Serializable {

  private static final long serialVersionUID = 2874825626697147072L;
  Long targetId;
  LocalDateTime partitionTs;
  PartitionUnit partitionUnit;
  int partitionSize;
  boolean partitioned;
  boolean bookmarked;

  String maxBookmark;

  public static PartitionBuilder builder() {
    return new PartitionBuilder();
  }

  public Partition(PartitionBuilder builder) {
    this.id = builder.id;
    this.targetId = builder.targetId;
    this.partitionTs = builder.partitionTs;
    this.partitionUnit = builder.partitionUnit;
    this.partitionSize = builder.partitionSize;
    this.partitioned = builder.partitioned;
    this.bookmarked = builder.bookmarked;
    this.maxBookmark = builder.maxBookmark;
  }

  public void setPartitioned(boolean partitioned) {
    this.partitioned = partitioned;
  }

  public void setKey() {
    StringBuilder res =
        new StringBuilder()
            .append(targetId)
            .append(partitionTs.toEpochSecond(ZoneOffset.UTC))
            .append(partitioned);
    long id = LongHashFunction.xx().hashChars(res);
    this.setId(id);
  }

  public Long getTargetId() {
    return targetId;
  }

  public void setTargetId(Long targetId) {
    this.targetId = targetId;
  }

  public LocalDateTime getPartitionTs() {
    return partitionTs;
  }

  public void setPartitionTs(LocalDateTime partitionTs) {
    this.partitionTs = partitionTs;
    ZonedDateTime partitionTsUTC = this.partitionTs.atZone(ZoneId.of("UTC"));
  }

  public boolean isPartitioned() {
    return partitioned;
  }

  public boolean isBookmarked() {
    return bookmarked;
  }

  public void setBookmarked(boolean bookmarked) {
    this.bookmarked = bookmarked;
  }

  public String getMaxBookmark() {
    return maxBookmark;
  }

  public void setMaxBookmark(String maxBookmark) {
    this.maxBookmark = maxBookmark;
  }

  public PartitionUnit getPartitionUnit() {
    return partitionUnit;
  }

  public void setPartitionUnit(PartitionUnit partitionUnit) {
    this.partitionUnit = partitionUnit;
  }

  public int getPartitionSize() {
    return partitionSize;
  }

  public void setPartitionSize(int partitionSize) {
    this.partitionSize = partitionSize;
  }

  @Override
  public int compareTo(Partition o) {
    return this.getId().compareTo(o.getId());
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof Partition)) {
      return false;
    }
    Partition partition = (Partition) o;
    return getPartitionSize() == partition.getPartitionSize()
        && isPartitioned() == partition.isPartitioned()
        && isBookmarked() == partition.isBookmarked()
        && Objects.equal(getTargetId(), partition.getTargetId())
        && Objects.equal(getPartitionTs(), partition.getPartitionTs())
        && getPartitionUnit() == partition.getPartitionUnit()
        && Objects.equal(getMaxBookmark(), partition.getMaxBookmark());
  }

  @Override
  public int hashCode() {
    return Objects.hashCode(
        getTargetId(),
        getPartitionTs(),
        getPartitionUnit(),
        getPartitionSize(),
        isPartitioned(),
        isBookmarked(),
        getMaxBookmark());
  }

  @Override
  public String toString() {
    return "Partition{"
        + "id="
        + id
        + ", targetId="
        + targetId
        + ", partitionTs="
        + partitionTs
        + ", partitionUnit="
        + partitionUnit
        + ", partitionSize="
        + partitionSize
        + ", partitioned="
        + partitioned
        + ", bookmarked="
        + bookmarked
        + ", maxBookmark='"
        + maxBookmark
        + '\''
        + '}';
  }
}

File: uhrwerk-metastore/src/main/kotlin/io/qimia/uhrwerk/common/metastore/model/Partition.kt
package io.qimia.uhrwerk.common.metastore.model

import java.sql.Timestamp
import java.time.LocalDateTime

data class Partition(
    var id: Long? = null,
    var tableKey: Long? = null,
    var targetKey: Long? = null,
    var partitionTs: LocalDateTime? = null,
    var partitionUnit: PartitionUnit? = null,
    var partitionSize: Int = 0,
    var partitioned: Boolean = false,
    var bookmarked: Boolean = false,
    var maxBookmark: String? = null,
    var partitionValues: Map<String, Any>? = null,
    var partitionPath: String? = null,
    var deactivatedTs: LocalDateTime? = null,
    var createdTs: Timestamp? = null,
    var updatedTs: Timestamp? = null
) : BaseModel {
    override fun id(id: Long?) {
        this.id = id
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Partition) return false
        if (tableKey != other.tableKey) return false
        if (targetKey != other.targetKey) return false
        if (partitionTs != other.partitionTs) return false
        if (partitionUnit != other.partitionUnit) return false
        if (partitionSize != other.partitionSize) return false
        if (partitioned != other.partitioned) return false
        if (bookmarked != other.bookmarked) return false
        if (maxBookmark != other.maxBookmark) return false
        if (!partitionValues.isNullOrEmpty()) {
            if (other.partitionValues.isNullOrEmpty()) return false
            if (partitionValues!! != other.partitionValues) return false
        } else if (!other.partitionValues.isNullOrEmpty()) return false
        if (deactivatedTs != other.deactivatedTs) return false
        return true
    }

    override fun hashCode(): Int {
        var result = tableKey?.hashCode() ?: 0
        result = 31 * result + (targetKey?.hashCode() ?: 0)
        result = 31 * result + (partitionTs?.hashCode() ?: 0)
        result = 31 * result + (partitionUnit?.hashCode() ?: 0)
        result = 31 * result + partitionSize
        result = 31 * result + partitioned.hashCode()
        result = 31 * result + bookmarked.hashCode()
        result = 31 * result + (maxBookmark?.hashCode() ?: 0)
        result = 31 * result + (partitionValues?.hashCode() ?: 0)
        result = 31 * result + (partitionPath?.hashCode() ?: 0)
        return result
    }

    override fun toString(): String {
        return "Partition(id=$id, tableKey=$tableKey, targetKey=$targetKey, partitionTs=$partitionTs, partitionUnit=$partitionUnit, partitionSize=$partitionSize, partitioned=$partitioned, bookmarked=$bookmarked, maxBookmark=$maxBookmark, partitionValues=$partitionValues, partitionPath=$partitionPath, deactivatedTs=$deactivatedTs, createdTs=$createdTs, updatedTs=$updatedTs)"
    }


}